### PR TITLE
Adding tests with RelyOnConnectionTimeoutsInsteadOfPollingRequestMaximumMessageProcessingTimeout turned on

### PR DESCRIPTION
**Pull request by sburmanoctopus**
[sc-59506]

# Background

We wanted some tests to prove that `RelyOnConnectionTimeoutsInsteadOfPollingRequestMaximumMessageProcessingTimeout` works in general.

We cannot just turn it on by default yet, as we want the 'off' path to be tested thoroughly. Mainly because primarily, we are still using in production with this setting off.

# How to review this PR

Quality :heavy_check_mark:

# Pre-requisites

- [ ] I have read [How we use GitHub Issues](https://github.com/OctopusDeploy/Issues/blob/master/docs/CONTRIBUTING.internal.md) for help deciding when and where it's appropriate to make an issue.
- [ ] I have considered informing or consulting the right people, according to the [ownership map](https://whimsical.com/ownership-map-NzbiD4HJyvhC9jNJNfS6TG).
- [ ] I have considered appropriate testing for my change.

### CODE REVIEW

File: source/Halibut.Tests/ExceptionContractFixture.cs [hunk 1]
```diff
@@ -61,6 +61,33 @@ public async Task WhenThePollingRequestMaximumMessageProcessingTimeoutIsReached_
             waitSemaphore.Release();
         }
 
+        [Test]
+        [LatestClientAndLatestServiceTestCases(testNetworkConditions: false, testListening: false)]
+        public async Task WhenThePollingRequestHasBegunTransfer_AndRelyingOnConnectionTimeoutsInsteadOfPollingRequestMaximumMessageProcessingTimeout_ThenTimeoutIsReached_AHalibutClientExceptionShouldBeThrown_Rename(ClientAndServiceTestCase clientAndServiceTestCase)
```
Comment (LukeButters) [inline source/Halibut.Tests/ExceptionContractFixture.cs:66]: _Rename?

File: source/Halibut.Tests/PollingServiceTimeoutsFixture.cs [hunk 2]
```diff
@@ -105,5 +133,46 @@ await using (var clientAndService = await clientAndServiceTestCase.CreateTestCas
                 }, TimeSpan.FromSeconds(30), Logger, CancellationToken);
             }
         }
+
+        [Test]
+        [LatestClientAndLatestServiceTestCases(testNetworkConditions: false, testListening: false)]
+        public async Task WhenThePollingRequestHasBegunTransfer_TheRequestShouldTimeout_AndTheTransferringPendingRequestCancelled(ClientAndServiceTestCase clientAndServiceTestCase)
```
Comment (LukeButters) [inline source/Halibut.Tests/PollingServiceTimeoutsFixture.cs:139]: missing big long name?

Comment (LukeButters) [reply source/Halibut.Tests/PollingServiceTimeoutsFixture.cs:139]: `TheRequestShouldTimeout` -> `AndWeTimeoutWaitigForTheResponse`

File: source/Halibut.Tests/PollingServiceTimeoutsFixture.cs [hunk 3]
```diff
@@ -105,5 +133,46 @@ await using (var clientAndService = await clientAndServiceTestCase.CreateTestCas
...
+        {
+            var halibutTimeoutsAndLimits = new HalibutTimeoutsAndLimitsForTestsBuilder().Build();
+            halibutTimeoutsAndLimits.PollingRequestQueueTimeout = TimeSpan.FromSeconds(5);
+            halibutTimeoutsAndLimits.TcpClientReceiveResponseTimeout = TimeSpan.FromSeconds(6);
+            halibutTimeoutsAndLimits.RelyOnConnectionTimeoutsInsteadOfPollingRequestMaximumMessageProcessingTimeout = true;
+
+            var waitSemaphore = new SemaphoreSlim(0, 1);
```
Comment (LukeButters) [inline source/Halibut.Tests/PollingServiceTimeoutsFixture.cs:146]: todo using

File: source/Halibut.Tests/PollingServiceTimeoutsFixture.cs [hunk 4]
```diff
@@ -105,5 +133,46 @@ await using (var clientAndService = await clientAndServiceTestCase.CreateTestCas
...
+                    .And.Message.Should().ContainAny(
+                        "Unable to read data from the transport connection: Connection timed out.",
+                        "Unable to read data from the transport connection: A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond");
+                stopwatch.Stop();
+
+                stopwatch.Elapsed.Should().BeLessThan(TimeSpan.FromSeconds(15), "Should have timed out quickly");
+                
+                Wait.UntilActionSucceeds(() =>
+                {
+                    connectionsObserver.ConnectionClosedCount.Should().Be(1, "Cancelling the PendingRequest will time out, causing the connection to be closed");
```
Comment (LukeButters) [inline source/Halibut.Tests/PollingServiceTimeoutsFixture.cs:170]: Maybe change the msg, we just want to see that when we timeout waiting for the response we kill that connection.

**Review by LukeButters — APPROVED** [review]
LGTM couple of suggestions

File: source/Halibut.Tests/ExceptionContractFixture.cs (hunk 1 above)
Comment (sburmanoctopus) [reply source/Halibut.Tests/ExceptionContractFixture.cs:66]: Fixed

File: source/Halibut.Tests/PollingServiceTimeoutsFixture.cs (hunk 2 above)
Comment (sburmanoctopus) [reply source/Halibut.Tests/PollingServiceTimeoutsFixture.cs:139]: Done

File: source/Halibut.Tests/PollingServiceTimeoutsFixture.cs (hunk 3 above)
Comment (sburmanoctopus) [reply source/Halibut.Tests/PollingServiceTimeoutsFixture.cs:146]: Done

**Pull request merged**